overviews/implement-api-bindings.md

```diff
@@ -0,0 +1,175 @@
+
+## API Transports
+
+Like with everything else, IPFS aims to be flexible regarding the API transports. Currently, the [go-ipfs](https://github.com/ipfs/go-ipfs) implementation supports both an in-process API and an HTTP api. More can be added easily, by mapping the API functions over a transport. (This is similar to how gRPC is also _mapped on top of transports_, like HTTP).
+
+Mapping to a transport involves leveraging the transport's features to express function calls. For example:
+
+### CLI API Transport
+
+In the commandline, IPFS uses a traditional flag and arg-based mapping, where:
+- the first arguments selects the command, as in git - e.g. `ipfs object get`
+- the flags specify options - e.g. `--enc=protobuf -q`
+- the rest are positional arguments - e.g. `ipfs object patch <hash1> add-link foo <hash2>`
+- files are specified by filename, or through stdin
+
+(NOTE: when go-ipfs runs the daemon, the CLI API is actually converted to HTTP calls. otherwise, they execute in the same process)
+
+### HTTP API Transport
+
+In HTTP, our API layering uses a REST-like mapping, where:
+- the URL path selects the command - e.g `/object/get`
+- the URL query string implements option arguments - e.g. `&enc=protobuf&q=true`
+- the URL query also implements positional arguments - e.g. `&arg=<hash1>&arg=add-link&arg=foo&arg=<hash2>`
+- the request body streams file data - reads files or stdin
+  - multiple streams are muxed with multipart (todo: add tar stream support)
+
+
+## API Commands
+
+There is a "standard IPFS API" with a set of commands, which we are documenting clearly soon. But this is not yet extracted into its own document. Perhaps -- as part of this API Bindings effort -- we can document it all. It is currently defined as "all the commands exposed by the go-ipfs implementation". You can see [a listing here](https://github.com/ipfs/go-ipfs/blob/916f987de2c35db71815b54bbb9a0a71df829838/core/commands/root.go#L82-L111), or by running `ipfs commands` locally. **The good news is: we should be able to easily write a program that outputs a markdown API specification!**
+
+(Note: the go-ipfs [commands library](https://github.com/ipfs/go-ipfs/tree/916f987de2c35db71815b54bbb9a0a71df829838/commands) also makes sure to keep the CLI and the HTTP API exactly in sync.)
+
+## Implementing bindings for the HTTP API
+
+As mentioned above, the API commands map to HTTP with:
+- the URL path selects the command - e.g `/object/get`
+- the URL query string implements option arguments - e.g. `&enc=protobuf&q=true`
+- the URL query also implements positional arguments - e.g. `&arg=<hash1>&arg=add-link&arg=foo&arg=<hash2>`
+- the request body streams file data - reads files or stdin
+  - multiple streams are muxed with multipart (todo: add tar stream support)
```

Comment:
So it is decided that we are moving towards tar instead of multipart? That would definitely solve a lot of vinyl madness :) //cc @dignifiedquire